crates/bevy_render/src/mesh/allocator.rs

```diff
@@ -0,0 +1,1025 @@
+                    None => {
+                        // We failed to insert one of the allocations that we
+                        // had before.
+                        continue 'grow;
+                    }
+                }
+            }
+
+            return Ok(slab_to_grow);
+        }
+    }
+}
+
+impl ElementLayout {
+    /// Creates an [`ElementLayout`] for mesh data of the given class (vertex or
+    /// index) with the given byte size.
+    fn new(class: ElementClass, size: u64) -> ElementLayout {
+        ElementLayout {
+            class,
+            size,
+            // Make sure that slot boundaries begin and end on
+            // `COPY_BUFFER_ALIGNMENT`-byte (4-byte) boundaries.
+            elements_per_slot: (COPY_BUFFER_ALIGNMENT / gcd(size, COPY_BUFFER_ALIGNMENT)) as u32,
+        }
+    }
+
+    fn slot_size(&self) -> u64 {
+        self.size * self.elements_per_slot as u64
+    }
+
+    /// Creates the appropriate [`ElementLayout`] for the given mesh's vertex
+    /// data.
+    fn vertex(
+        mesh_vertex_buffer_layouts: &mut MeshVertexBufferLayouts,
+        mesh: &Mesh,
+    ) -> ElementLayout {
+        let mesh_vertex_buffer_layout =
+            mesh.get_mesh_vertex_buffer_layout(mesh_vertex_buffer_layouts);
+        ElementLayout::new(
+            ElementClass::Vertex,
+            mesh_vertex_buffer_layout.0.layout().array_stride,
+        )
+    }
+
+    /// Creates the appropriate [`ElementLayout`] for the given mesh's index
+    /// data.
+    fn index(mesh: &Mesh) -> Option<ElementLayout> {
+        let size = match mesh.indices()? {
+            Indices::U16(_) => 2,
+            Indices::U32(_) => 4,
+        };
+        Some(ElementLayout::new(ElementClass::Index, size))
+    }
+}
+
+impl GeneralSlab {
+    /// Returns true if this slab is empty.
+    fn is_empty(&self) -> bool {
+        self.resident_allocations.is_empty() && self.pending_allocations.is_empty()
+    }
+}
+
+/// Returns the greatest common divisor of the two numbers.
+///
+/// <https://en.wikipedia.org/wiki/Euclidean_algorithm#Implementations>
+fn gcd(mut a: u64, mut b: u64) -> u64 {
+    while b != 0 {
+        let t = b;
+        b = a % b;
+        a = t;
+    }
+    a
+}
+
+/// Ensures that the size of a buffer is a multiple of the given alignment by
+/// padding it with zeroes if necessary.
+///
+/// If the buffer already has the required size, then this function doesn't
+/// allocate. Otherwise, it copies the buffer into a new one and writes the
+/// appropriate number of zeroes to the end.
+fn pad_to_alignment(buffer: &[u8], align: usize) -> Cow<[u8]> {
+    if buffer.len() % align == 0 {
+        return Cow::Borrowed(buffer);
+    }
+    let mut buffer = buffer.to_vec();
+    while buffer.len() % align != 0 {
+        buffer.push(0);
+    }
+    Cow::Owned(buffer)
+}
+
+/// Returns either `"vertex "` or `"index "`, depending on the type of buffer
```

Comment:
I'd rewrite this whole doc comment to something much simpler without the first paragraph. Maybe just like map BufferUsages to a string.